### DISCUSSION TITLE
Broadpahse not returning expected results

### DISCUSSION THREAD

**Discussion by rishavgmr** [NVIDIA-Omniverse/PhysX#251] [Q&A] [answered]
Hello all,

I am just starting to learn about PhysX 5.0 and have been trying out some of the features that will be relevant to my application. I need to use PhysX for collision detection and distance measurement.

For testing collision detection, I loaded a octomap file (.bt) by reading in each leaf voxel as a box primitive into a scene and then running a overlap scene query against a test sphere. This test returned expected results, specifically that the sphere was in fact in collision with one of the voxels/box primitives. Here is the code I used for that:

```
// includes

int main() {
    // PhysX setup including, Foundation, Pvd and Physics

    // Setup scene
    physx::PxSceneDesc scene_desc(physics->getTolerancesScale());
    scene_desc.gravity = physx::PxVec3(0.0f, -9.81f, 0.0f);
    physx::PxCpuDispatcher* gDispatcher = physx::PxDefaultCpuDispatcherCreate(2);
    scene_desc.cpuDispatcher = gDispatcher;
    scene_desc.filterShader = physx::PxDefaultSimulationFilterShader;
    physx::PxScene* scene = physics->createScene(scene_desc);

    std::string filename = "/workspaces/advanced-coliisions/data/octomap.bt";

    std::unique_ptr<octomap::OcTree> env_octree;
    env_octree = std::make_unique<octomap::OcTree>(filename);

    double box_side_length = env_octree->begin_leafs().getSize();
    physx::PxMaterial* material = physics->createMaterial(0.1f, 0.1f, 0.1f);
    physx::PxShape* voxel_shape =
        physics->createShape(physx::PxBoxGeometry(box_side_length / 2.0, box_side_length / 2.0, box_side_length / 2.0), *material);

    physx::PxAggregateFilterHint hint = physx::PxGetAggregateFilterHint(physx::PxAggregateType::eSTATIC, false);
    physx::PxAggregate* aggregate = physics->createAggregate(env_octree->getNumLeafNodes(), env_octree->getNumLeafNodes(), hint);

    double x, y, z;
    for (auto it = env_octree->begin_leafs(); it != env_octree->end_leafs(); ++it) {
        physx::PxRigidStatic* actor = physics->createRigidStatic(physx::PxTransform(physx::PxVec3(it.getX(), it.getY(), it.getZ())));
        x = it.getX();
        y = it.getY();
        z = it.getZ();

        actor->attachShape(*voxel_shape);
        aggregate->addActor(*actor);
    }
    scene->addAggregate(*aggregate);

    physx::PxOverlapBuffer hit;
    physx::PxSphereGeometry overlap_shape(0.15f);
    physx::PxQuat quat2(0.0f, 0.0f, 0.0f, 1.0f);
    physx::PxTransform transf2(-0.7, 0.400000, -0.005950, quat2);
    physx::PxTransform shape_pose(transf2);

    physx::PxQueryFilterData query_filter;
    query_filter.flags |= physx::PxQueryFlag::eANY_HIT; 

    bool overlap_result = scene->overlap(overlap_shape, shape_pose, hit, query_filter);
    std::cout << "Overlap result : " << hit.hasAnyHits() << std::endl;

    return 0;
}

```

Now I wanted to go to the next step which was distance computation for which I figured I would need broadphase as a first step to calculate which voxels are closest to the sphere. For that I wrote this:

```
// includes

int main() {
    // PhysX setup including, Foundation, Pvd and Physics

    // Setup scene
    std::string filename = "/workspaces/advanced-coliisions/data/octomap.bt";

    std::unique_ptr<octomap::OcTree> env_octree;
    env_octree = std::make_unique<octomap::OcTree>(filename);

    double box_side_length = env_octree->begin_leafs().getSize();
    physx::PxMaterial* material = physics->createMaterial(0.1f, 0.1f, 0.1f);
    physx::PxShape* voxel_shape =
        physics->createShape(physx::PxBoxGeometry(box_side_length / 2.0, box_side_length / 2.0, box_side_length / 2.0), *material);
    std::cout << "Box side length : " << box_side_length << std::endl;
    std::cin.get();

    physx::PxBroadPhaseDesc bp_desc(physx::PxBroadPhaseType::eABP);
    bp_desc.mDiscardStaticVsKinematic = false;
    std::cout << "Broadphase spec valid : " << bp_desc.isValid() << std::endl;
    physx::PxBroadPhase* broadphase = physx::PxCreateBroadPhase(bp_desc);
    physx::PxAABBManager* aabb_manager = physx::PxCreateAABBManager(*broadphase);

    int obj_idx = 0;
    for (auto it = env_octree->begin_leafs(); it != env_octree->end_leafs(); ++it) {
        physx::PxBounds3 bounds;
        physx::PxGeometryQuery::computeGeomBounds(bounds, physx::PxBoxGeometry(box_side_length / 2.0, box_side_length / 2.0, box_side_length / 2.0),
                                                  physx::PxTransform(physx::PxVec3(it.getX(), it.getY(), it.getZ())));
        aabb_manager->addObject(obj_idx, bounds, physx::PxGetBroadPhaseStaticFilterGroup());
        obj_idx++;
    }

    physx::PxSphereGeometry overlap_shape(0.15f);
    physx::PxQuat quat2(0.0f, 0.0f, 0.0f, 1.0f);
    physx::PxTransform transf2(-0.7, 0.400000, -0.005950, quat2);
    physx::PxTransform shape_pose(transf2);

    physx::PxBounds3 bounds;
    physx::PxGeometryQuery::computeGeomBounds(bounds, overlap_shape, shape_pose);
    aabb_manager->addObject(env_octree->getNumLeafNodes(), bounds, physx::PxGetBroadPhaseDynamicFilterGroup(env_octree->getNumLeafNodes()));

    physx::PxBroadPhaseResults results;
    aabb_manager->update(results);
    aabb_manager->fetchResults(results);

    std::cout << "Num pairs " << results.mNbCreatedPairs << " and " << results.mNbDeletedPairs << std::endl;

    for (int i = 0; i < results.mNbCreatedPairs; i++) {
        std::cout << "Possible collision between " << results.mCreatedPairs[i].mID0 << " and " << results.mCreatedPairs[i].mID1 << std::endl;
    }

    return 0;
}

``` 

As you can see the sphere is placed at the same location as the last one and has the same size. However, I always see the following output:

`Num pairs 0 and 0`

i.e. The number of created pairs and deleted pairs are both 0. Is there something wrong I am doing in the setup?

**Comment by PierreTerdiman** [comment]
Isn't it because of the duplicate updates?

```
    aabb_manager->update(results);
    aabb_manager->update(results);

```
The results are incremental so if nothing changes between the two updates, the number of "created pairs" will be zero indeed (it will be non-zero after the first update, not after the second one).

**Reply by rishavgmr** [reply]
Thank you for the response. I had added the second update call as a test. Even when I have just one update call, the output is still the same. I will go ahead and make the correction in the top post.

**Reply by PierreTerdiman** [reply]
Ok. I don't see anything obviously wrong in the rest of the code.

You could try to change the broadphase type (PxBroadPhaseType::eABP in your snippet) to see if it's a problem with a specific implementation, or with all of them. If it's with all of them, the issue is most likely in your setup.

Could you maybe just dump the data to a file and share it with us?

**Reply by rishavgmr** [reply]
Yes of course. That's what I thought as well. I did try with the other broadphase types anyway and the results were the same. I am attaching an archive containing the two source files (octomap_collision.cpp is the one that is working, octomap_broadphase is the second one that is not working as expected), the octomap.bt file containing the data to create the scene, and a Dockerfile to setup the environment if it becomes necessary. Let me know if you need anything else.

Thank you so much for the help!
[github_discussions_data.zip](https://github.com/NVIDIA-Omniverse/PhysX/files/14528737/github_discussions_data.zip)

**Reply by PierreTerdiman** [reply]
Hmmm I cannot really use this, the data is a binary file with an unknown format. I'd need all the octree files so that I can compile this, or a description of the format.

Could you just dump the data from that snippet? You know, just fopen / fwrite / fclose the binary data for the bounds and I could use that.

**Reply by rishavgmr** [reply]
Yes of course. I dumped the bounds data into this csv file. Each row described bounds for each voxel. The first row should explain what each column represents.

[bounds.csv](https://github.com/NVIDIA-Omniverse/PhysX/files/14529477/bounds.csv)

**Reply by PierreTerdiman** [reply]
Here is what I see when using this data. It is unclear if "dimensions" was the full size of a box or the half-size, but either way I'm getting some collisions. See attached file.

[SnippetStandaloneBroadphase.cpp.txt](https://github.com/NVIDIA-Omniverse/PhysX/files/14529776/SnippetStandaloneBroadphase.cpp.txt)

![image](https://github.com/NVIDIA-Omniverse/PhysX/assets/11311181/ab1592b4-f97b-4ab6-966d-6fd9708b0cf0)

**Reply by rishavgmr** [reply]
Dimensions are the full size of the box.

**Reply by PierreTerdiman** [reply]
Alright, in that case I am getting 2 overlaps and the screenshot looks like this. Are you able to use the code I attached to the previous post, to compare to what's happening in your initial code?

![image](https://github.com/NVIDIA-Omniverse/PhysX/assets/11311181/a11b4965-c6e3-4857-975e-00ff9de675bf)

**Reply by PierreTerdiman** [reply]
Here's an updated version that properly renders the sphere used for the query.

[SnippetStandaloneBroadphase.cpp.txt](https://github.com/NVIDIA-Omniverse/PhysX/files/14530240/SnippetStandaloneBroadphase.cpp.txt)


![image](https://github.com/NVIDIA-Omniverse/PhysX/assets/11311181/dd35bc55-5fe9-4a28-a469-e9072a6d82e7)

**Reply by rishavgmr** [reply]
Ok so I looked through the source file you attached. The only difference I saw was that I was not using the centerExtents function to create a "PxBounds3" object. In the original, I would simply use computeGeomBounds and add the resultant PxBounds3 object to the aabb_manager using addObject. But now I tried by creating a new PxBounds3 object using centerExtents and then adding to the aabb_manager and it seems to suddenly work! I am also seeing it overlap with 2 voxels.

Earlier:
```
physx::PxBounds3 bounds;
        physx::PxGeometryQuery::computeGeomBounds(bounds, physx::PxBoxGeometry(box_side_length / 2.0, box_side_length / 2.0, box_side_length / 2.0),
                                                  physx::PxTransform(physx::PxVec3(it.getX(), it.getY(), it.getZ())));
        aabb_manager->addObject(obj_idx, bounds, physx::PxGetBroadPhaseStaticFilterGroup());
        obj_idx++;
```

Now:
```
physx::PxBounds3 bounds;
        physx::PxGeometryQuery::computeGeomBounds(bounds, physx::PxBoxGeometry(box_side_length / 2.0, box_side_length / 2.0, box_side_length / 2.0),
                                                  physx::PxTransform(physx::PxVec3(it.getX(), it.getY(), it.getZ())));
        const physx::PxVec3 center(bounds.getCenter().x, bounds.getCenter().y, bounds.getCenter().z);
        const physx::PxVec3 extents(0.5 * bounds.getDimensions().x, 0.5 * bounds.getDimensions().y, 0.5 * bounds.getDimensions().z);
        physx::PxBounds3 bounds_new = physx::PxBounds3::centerExtents(center, extents);
        aabb_manager->addObject(obj_idx, bounds_new, physx::PxGetBroadPhaseStaticFilterGroup());
        obj_idx++;
```

Is this expected behavior? What is the reason that one works but the other doesn't?

**Reply by PierreTerdiman** [reply]
Sorry, no idea, that makes no sense at all :)

This:

```
const physx::PxVec3 center(bounds.getCenter().x, bounds.getCenter().y, bounds.getCenter().z);
const physx::PxVec3 extents(0.5 * bounds.getDimensions().x, 0.5 * bounds.getDimensions().y, 0.5 * bounds.getDimensions().z);

```
Should be the same as:

```
const physx::PxVec3 center = bounds.getCenter();
const physx::PxVec3 extents = bounds.getExtents();

```
And thus bounds should be equal to bounds_new. And the bounds you computed with "computeGeomBounds" are the ones you gave me in that Excel file, right? So they're the ones I used as well, and they should be correct.

Maybe you can assert that bounds.minimum == new_bounds.minimum, same for maximum, and see if anything gets triggered. But otherwise I can't explain this.

**Reply by PierreTerdiman** [reply]
Could it be some compiler bug maybe? Are you compiling in Release or Debug? (Try both).

If it's a compiler bug, maybe adding any kind of code between computeGeomBounds and addObject is enough to avoid it. In that case, try adding something else instead of the centerExtents call (maybe just a printf).

Sorry, running out of ideas. Maybe there is something obvious I don't see at this time (past midnight here). Will have another look tomorrow.

**Reply by rishavgmr** [reply, marked as answer]
Thank you so much for all the help. So I found the issue was that I was calling aabb_manager->fetchResults(results) after aabb_manager->update(results). I guess that replaces the results that the update call was providing as no voxel had moved, thus no new pairs were created. However, calling aabb_manager->fetchResults(results) after aabb_manager->update() [NOTE the empty function argument], gives the expected output.
This solves the problem.

Tldr: Don't call fetchResults after update(results). Only call fetchResults after update().

**Reply by PierreTerdiman** [reply]
Oh wow. I should have seen this. I think it shows that this API could be clearer, and perhaps the single-function-helper should have been named something more explicit like "updateAndFetchResults" instead of just "update". Taking a note of that.

**Reply by rishavgmr** [reply]
That will make the purpose of the function clearer.
